File: 02/c/02.c
#include <stdio.h>

int main() {
    int part1 = 0, part2 = 0;

    // both parts
    FILE *fp = fopen("input", "r");
    int a, b, c;
    char pwd[100];
    while (fscanf(fp, "%d-%d %c: %s\n", &a, &b, &c, &pwd) > 0) {
        int count = 0;
        for (char *ch = pwd; *ch; ++ch) *ch == c && ++count;
        if (a <= count && count <= b) ++part1;
        if ((pwd[a-1] == c) != (pwd[b-1] == c)) ++part2;
    }
    fclose(fp);

    printf("%d\n%d\n", part1, part2);
}

File: 02/hs/02.hs
#!/usr/bin/env runhaskell

parse :: String -> (Int, Int, Char, String)
parse line =
    (read a, read $ drop 1 b, head ch, pwd)
        where [rng, ch, pwd] = words line
              (a, b) = span (/='-') rng

part1 = length . filter part1'
    where part1' (a, b, ch, pwd) = between a b $ length $ filter (==ch) pwd
          between a b c = a <= c && c <= b

part2 = length . filter part2'
    where part2' (a, b, ch, pwd) = (pwd !! pred a == ch) /= (pwd !! pred b == ch)

main = do
    input <- map parse . lines <$> readFile "input"
    putStrLn . show $ part1 input
    putStrLn . show $ part2 input
